### DISCUSSION TITLE
How do I typeset from an XML data file?

### DISCUSSION THREAD

**Discussion by ctrlcctrlv** [sile-typesetter/sile#1461] [Q&A] [answered]
Let's say I have a bunch of structured content according to some XML schema

I've only ever tried SILE in TeX mode

How do I typeset XML?

Should I convert the XML to SILE TeX?

**Comment by alerque** [comment]
Your two choices are 

* convert the existing XML schema to SILE commands or
* write a class that registeres SILE commands to process each element of your XML schema.

If the data is a jumble of knick knacks doing the former with an XML preprocessor might make sense. Otherwise for most cases The latter approach is suggested. The existing docbook class is an example of that approach.

**Comment by Omikhleia** [comment, marked as answer]
No you do not have to convert XML to SIL-language.

SILE comes with XML parsing natively, though of course the elements need to be supported by appropriate commands. E.g. if you have a `<sometag>blah</sometag>` in your XML, you need a `sometag` command/environment to be provided and implement something.

The way I've done it is:
1. Creating a package with a command for each XML element
    - Some of them being ignored or passthru with the **xmltricks** package, so I only have to focus on important elements.
    - In some cases I had to make my own extended version of a passthru, e.g. to ignore spaces etc. in structural elements.
    - Regular commands for each elements that needs to do something
3. Creating a SIL "preamble" that selects a class, includes the above package as well as any other necessary packages, set fonts, main language, etc. (It's, other that that, an empty document, that will be "wrapped" around the XML source)
4. Of course, you could also eventually end up creating your own class loading all these packages, esp. if you need some class support for specific sectioning, page masters, etc. 
- run `sile -I mypreamble.sil mydoc.xml`

As of working examples, in my [repo](https://github.com/Omikhleia/omikhleia-sile-packages) --still on SILE 0.12.5-- you can possibly check
- The `preamble` folder, the teixxx class and packages (with a sample XML in a subset of XML TEI schema in the `examples` folder.
- Likewise, but in a `bible-segond-usfx` branch rather than `main`, a preamble for the Segond bible in French, and _very_ draft `packages/experimental/usfx.lua` for supporting the USFX tagset (and again, the raw XML content to process in the `examples/examples/segond`) -- Though a draft work-in-progress, this case might be a better introduction that the previous, as, at this point, the USFX stuff is far more easier than structured TEI. It does process the 1400+ pages of the Segond bible, though do not expect it to be very neat and complete.

N.B. This is, as far as I recall, the route described in the manual chapter devoted to XML processing with SILE. This does work without much pain, _as long as_ the XML element names don't overlap with existing commands from your supporting packages or from SILE's base classes... (That's a thing we'd need to address, sooner or later, but I was not in that case though didn't bother for now. Another thing we might want to address is that ill-named xmltricks packages, which could be improved IMHO -- but these are separate considerations anyhow).
